### PR TITLE
[Storage] Introduce support for namespace in state store 

### PR DESCRIPTION
**Pull request by sitalkedia**
## Motivation

First PR for introducing fine-grained state storage support. This diff introduces a concept of namespace in the state store - with only account blob based namespace support to start with.   The key to the state store for a namespace object can be `namespace_`+ `rawkey`. This way we can support different namespace objects within the same state store with the guarantee that there will not be any key collision. In order to support this, all account related concept has been moved out of the storage layer and instead the storage layer majorly understands the concept of `StateStoreKey` and `StateStoreValue`, which are arbitrary byte arrays. Please note that this is a major format change on the storage side and is not backward compatible - we need to wipe out the previous data from dev-net before rolling this out. 

The diff touches many files because of the extensive API change, however, important changes to be reviewed at

`types/src/state_store/state_store_key.rs`
`types/src/state_store/state_store_value.rs`
`storage/storage-interface/src/state_view.rs`
`storage/aptosdb/src/state_store/mod.rs`
`execution/executor/src/components/apply_chunk_output.rs`

## Test Plan

Modified existing UTs for new API interface.

### CODE REVIEW

File: state-sync/aptos-data-client/src/aptosnet/mod.rs [hunk 1]
```diff
@@ -323,7 +323,7 @@ impl AptosDataClient for AptosNetDataClient {
         version: u64,
         start_account_index: u64,
         end_account_index: u64,
-    ) -> Result<Response<AccountStatesChunkWithProof>> {
+    ) -> Result<Response<StateStoreValueChunkWithProof>> {
```
Comment (msmouse) [inline state-sync/aptos-data-client/src/aptosnet/mod.rs:326]: `StateChunkWifhProof` would do..

File: storage/aptosdb/src/aptosdb_test.rs [hunk 2]
```diff
@@ -538,13 +540,16 @@ fn verify_committed_transactions(
         assert_eq!(txn_output_list_with_proof.transactions_and_outputs.len(), 1);
 
         // Fetch and verify account states.
-        for (addr, expected_blob) in txn_to_commit.account_states() {
-            let account_state_with_proof = db
-                .get_account_state_with_proof(*addr, cur_ver, ledger_version)
+        for (state_store_key, state_store_value) in txn_to_commit.state_store_updates() {
+            let resource_value_with_proof = db
```
Comment (msmouse) [inline storage/aptosdb/src/aptosdb_test.rs:544]: "resource" in the naming seems irrelevant

File: storage/aptosdb/src/state_store/mod.rs [hunk 3]
```diff
@@ -148,20 +147,20 @@ impl StateStore {
         Ok(ret)
     }
 
-    pub fn get_account_count(&self, version: Version) -> Result<usize> {
+    pub fn get_leaf_count(&self, version: Version) -> Result<usize> {
```
Comment (msmouse) [inline storage/aptosdb/src/state_store/mod.rs:150]: The State store is not explicitly a "tree", so maybe "get_value_count", or just "get_count'

File: storage/aptosdb/src/state_store/state_store_test.rs [hunk 4]
```diff
@@ -30,13 +27,17 @@ fn put_account_state_set(
 ) -> HashValue {
     let mut cs = ChangeSet::new();
     let expected_new_leaves = account_state_set.len();
+    let value_state_set: HashMap<_, _> = account_state_set
```
Comment (msmouse) [inline storage/aptosdb/src/state_store/state_store_test.rs:30]: `state_value_set`? `value_set`? `state_set`?

File: storage/aptossum/src/main.rs [hunk 5]
```diff
@@ -65,7 +67,7 @@ struct AccountCmd {
     #[structopt(short, long, short)]
     address: AccountAddress,
     #[structopt(subcommand)]
-    account_op: AccountOp,
+    resource_op: AccountOp,
```
Comment (msmouse) [inline storage/aptossum/src/main.rs:70]: This rename seems unnecessary?

File: storage/jellyfish-merkle/src/lib.rs [hunk 6]
```diff
@@ -144,6 +145,8 @@ pub trait TestValue: Value + Arbitrary + std::fmt::Debug + Eq + PartialEq + 'sta
 // `TestValue` for `AccountStateBlob` here. Ideally the module that defines the specific value like
 // `AccountStateBlob` should import the `Value` trait and implement it there.
 impl Value for aptos_types::account_state_blob::AccountStateBlob {}
```
Comment (msmouse) [inline storage/jellyfish-merkle/src/lib.rs:147]: still needed?

File: storage/jellyfish-merkle/src/lib.rs [hunk 7]
```diff
@@ -265,7 +268,7 @@ where
     /// The batch version of `put_value_sets`.
     pub fn batch_put_value_sets(
         &self,
-        value_sets: Vec<Vec<(HashValue, V)>>,
+        value_sets: Vec<Vec<(HashValue, &V)>>,
```
Comment (msmouse) [inline storage/jellyfish-merkle/src/lib.rs:271]: awesome

File: storage/storage-client/src/lib.rs [hunk 8]
```diff
@@ -53,22 +52,15 @@ impl StorageClient {
         bcs::from_bytes(&result)?
     }
 
-    pub fn get_account_state_with_proof_by_version(
+    pub fn get_value_with_proof_by_version(
```
Comment (msmouse) [inline storage/storage-client/src/lib.rs:55]: outside of the state_db, a `value` does sound ambiguous, let's use `state_value`  (or `state_item` but you didn't use `item` anywhere, so forget it :) )

File: storage/storage-client/src/lib.rs [hunk 9]
```diff
@@ -53,22 +52,15 @@ impl StorageClient {
...
-    > {
-        self.request(StorageRequest::GetAccountStateWithProofByVersionRequest(
-            Box::new(GetAccountStateWithProofByVersionRequest::new(
-                address, version,
-            )),
-        ))
+    ) -> std::result::Result<(Option<StateStoreValue>, SparseMerkleProof<StateStoreValue>), Error>
+    {
+        self.request(StorageRequest::GetValueWithProofByVersionRequest(Box::new(
+            GetValueWithProofByVersionRequest::new(state_store_key, version),
```
Comment (msmouse) [inline storage/storage-client/src/lib.rs:62]: GetStateValueWithProofByVersionRequest

File: storage/storage-client/src/lib.rs [hunk 10]
```diff
@@ -88,16 +80,15 @@ impl StorageClient {
 }
 
 impl DbReader for StorageClient {
-    fn get_account_state_with_proof_by_version(
+    fn get_value_with_proof_by_version(
```
Comment (msmouse) [inline storage/storage-client/src/lib.rs:83]: ditto

File: storage/storage-client/src/lib.rs [hunk 11]
```diff
@@ -88,16 +80,15 @@ impl StorageClient {
...
         &self,
-        address: AccountAddress,
+        resource_key: StateStoreKey,
```
Comment (msmouse) [inline storage/storage-client/src/lib.rs:85]: "resource" is ambiguous if not confusing

File: storage/storage-interface/src/lib.rs [hunk 12]
```diff
@@ -315,10 +320,7 @@ pub trait DbReader: Send + Sync {
     ///
     /// [`AptosDB::get_latest_account_state`]:
     /// ../aptosdb/struct.AptosDB.html#method.get_latest_account_state
-    fn get_latest_account_state(
-        &self,
-        address: AccountAddress,
-    ) -> Result<Option<AccountStateBlob>> {
+    fn get_latest_value(&self, state_store_key: StateStoreKey) -> Result<Option<StateStoreValue>> {
```
Comment (msmouse) [inline storage/storage-interface/src/lib.rs:323]: mention "state"

File: storage/storage-interface/src/lib.rs [hunk 13]
```diff
@@ -392,12 +394,12 @@ pub trait DbReader: Send + Sync {
 
     /// Returns the account state corresponding to the given version and account address with proof
     /// based on `ledger_version`
-    fn get_account_state_with_proof(
+    fn get_value_with_proof(
```
Comment (msmouse) [inline storage/storage-interface/src/lib.rs:397]: ditto

File: storage/storage-interface/src/lib.rs [hunk 14]
```diff
@@ -409,14 +411,11 @@ pub trait DbReader: Send + Sync {
     // ../aptosdb/struct.AptosDB.html#method.get_account_state_with_proof_by_version
     //
     // This is used by diem core (executor) internally.
-    fn get_account_state_with_proof_by_version(
+    fn get_value_with_proof_by_version(
```
Comment (msmouse) [inline storage/storage-interface/src/lib.rs:414]: ditto

File: storage/storage-interface/src/state_view.rs [hunk 15]
```diff
@@ -171,16 +177,18 @@ impl StateView for VerifiedStateView {
...
-                    .insert(address_hash, proof);
-
-                blob
+                self.state_proof_cache.write().insert(key_hash, proof);
+                state_store_value
             }
         };
 
+        // Hack: Convert the state store value to account blob option as that is the
+        // only type of state resource we support for now. This needs to change once we start
```
Comment (msmouse) [inline storage/storage-interface/src/state_view.rs:186]: don't use "state resource" 🤣

File: types/src/account_state.rs [hunk 16]
```diff
@@ -337,6 +339,38 @@ impl fmt::Debug for AccountState {
...
+    }
+}
+
+impl TryFrom<&mut StateStoreValue> for AccountState {
+    type Error = Error;
+
+    fn try_from(state_store_value: &mut StateStoreValue) -> Result<Self> {
+        AccountState::try_from(&state_store_value.bytes)
+    }
+}
```
Comment (msmouse) [inline types/src/account_state.rs:356]: Is this necessary?

**Review by msmouse — COMMENTED** [review]
Beautiful. The scope of change is well controlled. Just nits for naming:

1. `state_store_key` seems redundant vs `state_key`, similar for `StateStoreKey`, etc. I don't feel strong about this.
2. on the state store "state" is implied, but not on the `AptosDb` or the `DbReader`, so instead of `get_latest_value()`, we should use `get_latest_state_value()`.
3. "state resource" is confusing because we have the account resource and account module concepts, let's try not to use this concept.

File: storage/aptossum/src/main.rs (hunk 5 above)
Comment (sitalkedia) [reply storage/aptossum/src/main.rs:70]: Good catch.

File: storage/storage-interface/src/lib.rs (hunk 12 above)
Comment (sitalkedia) [reply storage/storage-interface/src/lib.rs:323]: Done!

File: storage/storage-interface/src/state_view.rs (hunk 15 above)
Comment (sitalkedia) [reply storage/storage-interface/src/state_view.rs:186]: Yes naming is hard :)

File: types/src/account_state.rs (hunk 16 above)
Comment (sitalkedia) [reply types/src/account_state.rs:356]: Yes there are some tests still generating arbitrary account state, so this is needed.

**Review by msmouse — APPROVED** [review]

**Comment by sitalkedia** [comment]
@msmouse - I checked the failed tests and it seems like these are due to the new hash value of the keys creating different number of intermediate nodes as compared to the previous hashing scheme, I have updated the test. Please let me know your thoughts - https://github.com/aptos-labs/aptos-core/pull/324/commits/cacb3f2add7bc85ffeba942511578d3acf19fb5e

**Comment by msmouse** [comment]
@sitalkedia  Makes sense! (And it's a tricky one!)

**Comment by msmouse** [comment]
I think we can squash the two commits before landing this one.
